### PR TITLE
Added PrependEventStream

### PR DESCRIPTION
**Pull request by JordanMartinez**

### CODE REVIEW

**Comment by JordanMartinez** [comment]
That test didn't fail for me.

**Comment by JordanMartinez** [comment]
I just want to clarify what prepend does: 
- My idea was that, until `stream` actually emits a value, the Observer who subscribes to the prepended stream will immediately be notified of the initial value. Once `stream` emits a real value, then any additional observer will be notified of the real value and not the initial value. So here's the process:
 1. Stream has not emitted its first event
 2. Observer A subscribes to Stream; Observer A is notified to `initial` event.
 3. Stream now emits its first event
 4. Observer A is notified of first event.
 5. Observer B subscribes to Stream; **Observer B is notified of first event _not_ `initial` event.** 
- The test you wrote makes me think you have a slightly different idea: an Observer will be notified of the initial value regardless of whether the real value has been emitted already. So, if `stream` has emitted a real event and then something new subscribes to it, that new observer will still receive the initial value _and_ the emitted real value. So here's my understanding of that process:
 1. Stream has not emitted its first event
 2. Observer A subscribes to Stream; Observer A is notified of the `initial` event.
 3. Stream now emits its first event
 4. Observer A is notified of first event.
 5. Observer B subscribes to Stream; **Observer B is notified of `initial` event, _and then_ first event.**

**Comment by JordanMartinez** [comment]
With that clarification, I think the EventStream I'm trying to write would be better named `InitialEventEventStream` (I don't like the repetition of Event in the name, so that should be changed), one where the EventStream only notifies every new subscriber of the initial event until the stream emits a real value, whereas the one that I think you are thinking of would be the true `PrependEventStream`, one where no matter when one subscribes, it will be notified of the initial event before the event that was emitted most recently.

**Comment by TomasMikula** [comment]
No, I share your idea of what `prepend` should do.

**Comment by JordanMartinez** [comment]
Ok. So shouldn't the assertion in the test you wrote be:
````java
assertEquals(Arrays.asList(1), emitted);
````
because the prepended event is overridden immediately?

**Comment by TomasMikula** [comment]
The first subscriber should get the initial value, shouldn't it?

**Comment by JordanMartinez** [comment]
Mm... I don't think so because the reason why we wanted to write this kind of EventStream (going back to the workaround in the wiki page) was to only provide an initial event until the stream emitted a real one.

**Comment by TomasMikula** [comment]
Such behavior makes sense as well. I think, however, that an extra event in that situation should not matter, because the user couldn't know there would be one immediately (if they did, they would not need to use prepend).

**Comment by JordanMartinez** [comment]
If I'm understanding you correctly, a developer would only use `prepend` when they have no way of knowing or controlling what the first event will be. So, in that situation, the extra event wouldn't create a problem. If they did know or were able to control the first event, they wouldn't use it.

In that case, I agree with you.

**Comment by JordanMartinez** [comment]
Still, it might help to write a warning in the javadoc just so some developer doesn't get an unexpected surprise if they happened to do that accidentally.

**Comment by TomasMikula** [comment]
When you hold and `EventStream<T>`, you don't know whether it will produce a value immediately upon subscription. Since you don't know, and you need one, you use `prepend` to make sure it does emit a value immediately. In the current implementation, it can then happen that it emits two values immediately. Suppressing the prepended value in such a case, as you suggest, would only by an optimization, but should not break anyone.

**Comment by JordanMartinez** [comment]
Yeah.

**Comment by TomasMikula** [comment]
I think both versions are equally meaningful, and for the purpose we have in mind are equivalent. If we go with your version, maybe we should call it something else. Even the current version does a little more than "prepend". It also re-emits the latest value to following subscribers.

**Comment by TomasMikula** [comment]
Here's what your version could look like

```java
    private T latestEvent = null;
    private boolean firstObserver = true;
    private boolean emitted = false;

    @Override
    protected void newObserver(Consumer<? super T> observer) {
        if(firstObserver) {
            firstObserver = false;
            if(!emitted) {
            	observer.accept(initial);
            }
        } else {
            observer.accept(latestEvent);
        }
    }

    @Override
    protected final Subscription observeInputs() {
        firstObserver = true;
        emitted = false;
        latestEvent = initial;
        return input.subscribe(x -> {
            latestEvent = x;
            emitted = true;
            emit(x);
        });
    }
```

**Comment by JordanMartinez** [comment]
Aye. So, both ideas should be implemented and the current form of prepend should be optimized to stop that double emission?

**Comment by TomasMikula** [comment]
I don't think we should include both, unless we have different use-cases for them. So far we have one use case, and either version works for it.

**Comment by JordanMartinez** [comment]
Ok.
I think the advantage of the idea I proposed is that it better reflects the use case we have. However, I don't know how best to name it.
With prepend in its current form, it could be used wrongly because it does more than just 'prepend.'

**Comment by JordanMartinez** [comment]
Just fyi, I'm going to sleep now, so don't wait on me for any answers. ☺
I'll respond tomorrow.

**Comment by TomasMikula** [comment]
:+1:

**Comment by JordanMartinez** [comment]
So... any new thoughts?

**Comment by TomasMikula** [comment]
About the name? I haven't thought too hard about it yet :)

**Comment by JordanMartinez** [comment]
:smile: 
The function of the new EventStream is to emit an Event immediately whenever anything subscribes to it. If the original stream has already emitted an event, then the latest event is emitted; if it hasn't, then initial is emitted.
So, some of my ideas for the name of the `EventStream` method and the returned EventStream's object's class name are:
- `untilFirst()` & `UntilFirstEventStream` (already shown)
 - `untilFirstEmission()` & `UntilFirstEmissionEventStream` (variant)
- `emitImmediately()` & `EmitLatestImmediatelyEventStream` (still doesn't cover the idea of only emitting initial until the stream actually emits one.
 - `emitImmediately()` & `ImmediateEventStream` (I think I actually like this one better than `UntilFirstEventStream`)
- `withInitialEvent()` & `WithInitialEventStream`

**Comment by TomasMikula** [comment]
Of these, I think I like `withInitialEvent()` the best. Or maybe would `withDefaultEvent()` be better, because it better captures the fact that the default event might not be emitted at all, if the underlying stream emits immediately.

**Comment by JordanMartinez** [comment]
I think `withDefaultEvent()` sounds better. It has the idea that, if there is nothing to emit, then it _defaults_ to emitting the given value.

File: reactfx/src/main/java/org/reactfx/DefaultEventStream.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package org.reactfx;
+
+import java.util.function.Consumer;
+
+public class DefaultEventStream<T> extends EventStreamBase<T> {
```
Comment (TomasMikula) [inline reactfx/src/main/java/org/reactfx/DefaultEventStream.java]: Doesn't need to be public, right?

Comment (JordanMartinez) [inline reactfx/src/main/java/org/reactfx/DefaultEventStream.java]: Right. I forgot to check the other classes that you wrote as an example for this area.
It's fixed in the next commit.

**Comment by TomasMikula** [comment]
:+1:

**Comment by JordanMartinez** [comment]
I've updated Example 1 in the wiki to use DefaultEventStream.

**Pull request merged**